Sentence checker

DLu opened the pull request:
And now, a reading from the book of [Code Style](https://docs.ros.org/en/rolling/The-ROS2-Project/Contributing/Code-Style-Language-Versions.html#markdown-restructured-text-docblocks)

> Each sentence must start on a new line.

@clalancette is a real stickler for this, so I made a linter. Unfortunately, there are lots of problems, some of which are the linter, and some of which are the content.

Happy to have help. I can't devote a ton more time to this, but I still wanted the code to be out there.

DLu committed:
Sentence Checker

DLu committed:
Some edits

DLu commented:
Very important note: I resisted the urge to call this ["Edgar Jr."](http://www.hrwiki.org/wiki/Edgar) 

https://www.youtube.com/watch?v=NW14YWw3h1E

DLu committed:
Update About-Executors.rst

fujitatomoya commented:
@DLu 

> Unfortunately, there are lots of problems, some of which are the linter, and some of which are the content.

i just got interested in this. can you rephrase this? are you saying there are many false alarms?

DLu committed:
Merge remote-tracking branch 'upstream/rolling' into sentence_checker

DLu commented:
Taking a look at the current results on my machine, it seems like mostly the output is true positives, with a handful of false positives mixed in. The false positives include: 
 * Numbered lists
 * Day abbreviations
 * People who use punctuation in their given names. Particularly ironic, coming from me, don't you think @nuclearsandwich ?

Current error count for the whole repo: 624

DLu committed:
More sentence breaks

fujitatomoya commented:
@DLu sounds good, if there is anything i can help, please let me know. i will take a look at the PR.

DLu committed:
More cleanup

DLu commented:
Down to 285

fujitatomoya commented:
ninja quick 🥷 amazing 💯

cottsay commented:
I haven't dug into this, but it might be good to actually parse the RST so we don't need to worry about handling the syntax. Example of where we're already parsing RST: https://github.com/ros-infrastructure/catkin_pkg/blob/master/src/catkin_pkg/changelog.py

fujitatomoya commented:
@Mergifyio rebase

fujitatomoya commented:
@DLu can you rebase this against rolling?

DLu committed:
Merge remote-tracking branch 'upstream/rolling' into sentence_checker

DLu committed:
More sentences

fujitatomoya committed:
fix group tag indent for ROSCon 2023 and 2024 contents. (#5016)

Signed-off-by: Tomoya Fujita <Tomoya.Fujita@sony.com>

christophebedard committed:
Exclude literals parts from bold parts in 'Ament Lint CLI Utilities' (#5024)

You can't do nested inline markup in RST:

**Can't do ``this`` in RST.**

The simplest solution is to just exclude the ``literal`` part.

Also, do some other minor fixes.

Signed-off-by: Christophe Bedard <christophe.bedard@apex.ai>

fujitatomoya committed:
remove --break-system-packages and address docker build warnings. (#5021)

* remove --break-system-packages and address docker build warnings.

Signed-off-by: Tomoya Fujita <Tomoya.Fujita@sony.com>

* add github workflow to test Dockerfile.

Signed-off-by: Tomoya Fujita <Tomoya.Fujita@sony.com>

* Revert "add github workflow to test Dockerfile."

This reverts commit 8686f18e6b931fad404fcfbaf0afbf79d218c0a9.

Signed-off-by: Tomoya Fujita <Tomoya.Fujita@sony.com>

---------

Signed-off-by: Tomoya Fujita <Tomoya.Fujita@sony.com>

DLu committed:
Fix smart quotes

DLu committed:
Other sentence formatting (#5031)

* Add second dot to abbreviations
* Remove some random periods
* Escape some periods
* Convert some links
* Add formatted block

DLu committed:
More

fujitatomoya committed:
Enable github/workflow to verify Dockerfile. (#5035)

Signed-off-by: Tomoya Fujita <Tomoya.Fujita@sony.com>

NickTziaros committed:
Update Using-Node-Interfaces-Template-Class.rst (#5038)

Added missing ``Output:`` before the terminal output for consistency reasons

Signed-off-by: Nikos Tziaros <33639811+NickTziaros@users.noreply.github.com>

NickTziaros committed:
added Publishing-Messages-Using-YAML-Files.rst and edited Intermediat… (#5034)

* added Publishing-Messages-Using-YAML-Files.rst and edited Intermediate.rst

* Update source/Tutorials/Intermediate/Publishing-Messages-Using-YAML-Files.rst

Co-authored-by: Tomoya Fujita <Tomoya.Fujita@sony.com>
Signed-off-by: Nikos Tziaros <33639811+NickTziaros@users.noreply.github.com>

* Update source/Tutorials/Intermediate/Publishing-Messages-Using-YAML-Files.rst

Co-authored-by: Tomoya Fujita <Tomoya.Fujita@sony.com>
Signed-off-by: Nikos Tziaros <33639811+NickTziaros@users.noreply.github.com>

* Update source/Tutorials/Intermediate/Publishing-Messages-Using-YAML-Files.rst

Signed-off-by: Nikos Tziaros <33639811+NickTziaros@users.noreply.github.com>

* Update source/Tutorials/Intermediate/Publishing-Messages-Using-YAML-Files.rst

Co-authored-by: Alejandro Hernández Cordero <ahcorde@gmail.com>
Signed-off-by: Nikos Tziaros <33639811+NickTziaros@users.noreply.github.com>

* Update source/Tutorials/Intermediate/Publishing-Messages-Using-YAML-Files.rst

Co-authored-by: Alejandro Hernández Cordero <ahcorde@gmail.com>
Signed-off-by: Nikos Tziaros <33639811+NickTziaros@users.noreply.github.com>

* Update source/Tutorials/Intermediate/Publishing-Messages-Using-YAML-Files.rst

Co-authored-by: Alejandro Hernández Cordero <ahcorde@gmail.com>
Signed-off-by: Nikos Tziaros <33639811+NickTziaros@users.noreply.github.com>

* Update source/Tutorials/Intermediate.rst

Thanks for pointing it out, I didnt notice the extention beeing there

Co-authored-by: Christophe Bedard <bedard.christophe@gmail.com>
Signed-off-by: Nikos Tziaros <33639811+NickTziaros@users.noreply.github.com>

* Update source/Tutorials/Intermediate/Publishing-Messages-Using-YAML-Files.rst

Co-authored-by: Christophe Bedard <bedard.christophe@gmail.com>
Signed-off-by: Nikos Tziaros <33639811+NickTziaros@users.noreply.github.com>

---------

Signed-off-by: Nikos Tziaros <33639811+NickTziaros@users.noreply.github.com>
Co-authored-by: Tomoya Fujita <Tomoya.Fujita@sony.com>
Co-authored-by: Alejandro Hernández Cordero <ahcorde@gmail.com>
Co-authored-by: Christophe Bedard <bedard.christophe@gmail.com>

DLu committed:
Clean up sentences II (white space changes only) (#5017)

DLu committed:
more bits

DLu commented:
Note: This PR has been split up. 
* The functionality is going in this repo: https://github.com/DLu/sphinx_tamer
* The content fixes are going in other PRs, such as #5001 #5017 #5030 and #5031 (and possibly more in the future)

DLu committed:
Merge remote-tracking branch 'upstream/rolling' into sentence_checker

DLu committed:
Some rollbacks

DLu committed:
Merge remote-tracking branch 'upstream/rolling' into sentence_checker

DLu committed:
Ignore

DLu commented:
@fujitatomoya My sentence checker is now checked in here: https://github.com/DLu/sphinx_tamer

At some point, I could reintegrate it with this PR, or we can work to integrate it into a GitHub check. Which do you think makes the most sense?

fujitatomoya commented:
@DLu sorry for being late back to you, i was on the vacation last week.

i am trying to catch up with you.

> At some point, I could reintegrate it with this PR, or we can work to integrate it into a GitHub check.

these are different implementation for me, can be in the same PR though.

what i am not really sure is,

since sphinx_tamer isn't available on the Python Package Index (PyPI), we'll need to install it directly from its GitHub repository, right? and the call it from here as sphinx extension via `make test` or another make target?
once this is clear, i think having it on github action is pretty straight forward.
could you enlighten me this design a bit?

DLu committed:
Use sphinx_tamer

DLu committed:
Merge remote-tracking branch 'upstream/rolling' into sentence_checker

DLu committed:
More Sphinx Config

DLu committed:
Updated checker using subprocess

sloretz commented on plugins/ros_checkers.py:
```diff
@@ -0,0 +1,16 @@
+
+@checker('.rst', '.md')
+def check_sentence_count(file, lines, options=None):
+    output = subprocess.check_output(['sphinx_sentence_scan_single', '.', file])
```
Where does `sphinx_sentence_scan_single` come from? It looks like the CI job is unable to find it

```
./sphinx-lint-with-ros source
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/usr/lib/python3.10/multiprocessing/pool.py", line [12](https://github.com/ros2/ros2_documentation/actions/runs/13578885182/job/37961093449?pr=4592#step:5:13)5, in worker
    result = (True, func(*args, **kwds))
  File "/home/runner/.local/lib/python3.10/site-packages/sphinxlint/cli.py", line 174, in _check_file
    return check_file(*todo)
  File "/home/runner/.local/lib/python3.10/site-packages/sphinxlint/sphinxlint.py", line 66, in check_file
    return check_text(filename, text, checkers, options)
  File "/home/runner/.local/lib/python3.10/site-packages/sphinxlint/sphinxlint.py", line 45, in check_text
    for lno, msg in check(
  File "/home/runner/work/ros2_documentation/ros2_documentation/plugins/ros_checkers.py", line 9, in check_sentence_count
    output = subprocess.check_output(['sphinx_sentence_scan_single', '.', file])
  File "/usr/lib/python3.10/subprocess.py", line 421, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.10/subprocess.py", line 503, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/usr/lib/python3.10/subprocess.py", line 971, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib/python3.10/subprocess.py", line [18](https://github.com/ros2/ros2_documentation/actions/runs/13578885182/job/37961093449?pr=4592#step:5:19)63, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'sphinx_sentence_scan_single'
"""
```

DLu replied:
My apologies for that, I forgot to push my changes to `sphinx_tamer` before pushing this branch. I've made a new release and hope the CI will work this time.

DLu committed:
Merge remote-tracking branch 'upstream/rolling' into sentence_checker

DLu commented:
Worth noting that this new version with `sphinx_sentence_scan_single` is a semi-hacky backup. I intended to run the Python directly instead of using `subprocess`, but Sphinx is a bit particular about how it runs as a library. 

[Here is the "clean" direct-Python version of the checker](https://github.com/DLu/ros2_documentation/blob/1a6c15ff53b51e9ff2631623fb7fde21f1abc303/plugins/ros_checkers.py#L31-L56)

However, this interacts with `sphinx_sitemap_ros.py` in a way that is currently beyond my understanding of Sphinx. 

```bash
! ros2_documentation/ > make lint
./sphinx-lint-with-ros source
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/sphinx/events.py", line 98, in emit
    results.append(listener.handler(self.app, *args))
  File "/home/dlu/OpenSource/ros2_documentation/plugins/sphinx_sitemap_ros.py", line 112, in record_builder_type
    builder.env.app.sitemap_links = Manager().Queue()
  File "/usr/lib/python3.10/multiprocessing/context.py", line 57, in Manager
    m.start()
  File "/usr/lib/python3.10/multiprocessing/managers.py", line 562, in start
    self._process.start()
  File "/usr/lib/python3.10/multiprocessing/process.py", line 118, in start
    assert not _current_process._config.get('daemon'), \
AssertionError: daemonic processes are not allowed to have children

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/lib/python3.10/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/usr/local/lib/python3.10/dist-packages/sphinxlint/cli.py", line 174, in _check_file
    return check_file(*todo)
  File "/usr/local/lib/python3.10/dist-packages/sphinxlint/sphinxlint.py", line 66, in check_file
    return check_text(filename, text, checkers, options)
  File "/usr/local/lib/python3.10/dist-packages/sphinxlint/sphinxlint.py", line 45, in check_text
    for lno, msg in check(
  File "/home/dlu/OpenSource/ros2_documentation/plugins/ros_checkers.py", line 34, in check_sentence_count
    settings, reporter, parser = load_sphinx(
  File "/home/dlu/Projects/sphinx_tamer/src/sphinx_tamer/sphinx_wrapper.py", line 47, in load_sphinx
    app = Sphinx(srcdir=src_path,
  File "/usr/local/lib/python3.10/dist-packages/sphinx/application.py", line 292, in __init__
    self._init_builder()
  File "/usr/local/lib/python3.10/dist-packages/sphinx/application.py", line 366, in _init_builder
    self.events.emit('builder-inited')
  File "/usr/local/lib/python3.10/dist-packages/sphinx/events.py", line 109, in emit
    raise ExtensionError(
sphinx.errors.ExtensionError: Handler <function record_builder_type at 0x7426fd50d900> for event 'builder-inited' threw an exception (exception: daemonic processes are not allowed to have children)
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/dlu/OpenSource/ros2_documentation/./sphinx-lint-with-ros", line 7, in <module>
    main(sys.argv)
  File "/usr/local/lib/python3.10/dist-packages/sphinxlint/cli.py", line 238, in main
    count = print_errors(
  File "/usr/local/lib/python3.10/dist-packages/sphinxlint/cli.py", line 199, in print_errors
    for error in errors:
  File "/usr/local/lib/python3.10/dist-packages/sphinxlint/cli.py", line 180, in sort_errors
    for results in results:
  File "/usr/lib/python3.10/multiprocessing/pool.py", line 873, in next
    raise value
sphinx.errors.ExtensionError: Handler <function record_builder_type at 0x7426fd50d900> for event 'builder-inited' threw an exception (exception: daemonic processes are not allowed to have children)
make: *** [Makefile:26: lint] Error 1
```

If the sitemap file is not there, it does not have this problem.

DLu committed:
Merge remote-tracking branch 'upstream/rolling' into sentence_checker

DLu commented:
Didn't realize I left this in draft state. Conflicts are resolved, ready for review.

sloretz commented on sphinx-lint-with-ros:
```diff
@@ -0,0 +1,7 @@
+import sys
+from sphinxlint.cli import main
+
+main(sys.argv)
```
```suggestion
sys.exit(main(sys.argv))
```

The checker returns errors, but the CI job passes. I think the fix is to call `sys.exit()` with [the return code from `sphinxlint.cli.main`](https://github.com/sphinx-contrib/sphinx-lint/blob/b1bec138033385653f42606a95af79a184dd08a5/sphinxlint/cli.py#L244)

https://github.com/ros2/ros2_documentation/actions/runs/13792027342/job/38574144923?pr=4592

```
./sphinx-lint-with-ros source
source/Tutorials/Intermediate/URDF/Using-URDF-with-Robot-State-Publisher-cpp.rst:307: Each sentence must start on a new line. Break between "put the r2d2." and "rviz into ``install``" (sentence-count)
source/Tutorials/Intermediate/URDF/Using-URDF-with-Robot-State-Publisher-cpp.rst:358: Each sentence must start on a new line. Break between "where the r2d2." and "rviz stored." (sentence-count)
source/Installation/Alternatives/Windows-Development-Setup.rst:51: Each sentence must start on a new line. Break between "limitation at https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation?" and "tabs=registry." (sentence-count)
```

sloretz committed:
Make sphinx-lint-with-ros exit with non-zero error codes

Signed-off-by: Shane Loretz <shane.loretz@gmail.com>

sloretz committed:
Wrap r2d2.rviz in backticks

Signed-off-by: Shane Loretz <sloretz@intrinsic.ai>

sloretz committed:
Wrap URL in RST link

Signed-off-by: Shane Loretz <sloretz@intrinsic.ai>

sloretz reviewed (approved):
LGTM! I'll ask for one more review since I made a few small changes.

christophebedard commented on source/The-ROS2-Project/Contributing/Code-Style-Language-Versions.rst:
```diff
@@ -376,16 +376,34 @@ Since there is not an existing CMake style guide we will define our own:
+
+* Each sentence can optionally be wrapped to keep each line short.
+* The lines should not have any trailing white spaces.
+
+Markdown or RST
+^^^^^^^^^^^^^^^
+* Each section title should be preceded by one empty line and succeeded by one empty line.
```
> Each section title should be preceded by one empty line and succeeded by one empty line.

Awkward :stuck_out_tongue_winking_eye: 

```suggestion

Markdown or RST
^^^^^^^^^^^^^^^

* Each section title should be preceded by one empty line and succeeded by one empty line.
```

christophebedard commented on source/The-ROS2-Project/Contributing/Code-Style-Language-Versions.rst:
```diff
@@ -376,16 +376,34 @@ Since there is not an existing CMake style guide we will define our own:
 The following rules to format text is intended to increase readability as well as versioning.
 
-* *[.md, .rst only]* Each section title should be preceded by one empty line and succeeded by one empty line.
+Any Doc Type
+^^^^^^^^^^^^
+* Each sentence must start on a new line.
```
```suggestion
Any Doc Type
^^^^^^^^^^^^

* Each sentence must start on a new line.
```

christophebedard reviewed (commented):
Minor comments.

sloretz committed:
Add empty line after section title

Co-authored-by: Christophe Bedard <bedard.christophe@gmail.com>
Signed-off-by: Shane Loretz <shane.loretz@gmail.com>

sloretz committed:
Add empty line after section title

Co-authored-by: Christophe Bedard <bedard.christophe@gmail.com>
Signed-off-by: Shane Loretz <shane.loretz@gmail.com>

nuclearsandwich reviewed (approved):
It's a huge level-up to have this as an automated check because this style requirement is motivated by making change review using textual diff tools much more natural.

Helping contributors check the constraint will save their time and maintainers'!

christophebedard commented:
We intend to backport this and fix the merge conflicts + linter issues, right?

fujitatomoya commented on requirements.txt:
```diff
@@ -9,3 +9,4 @@ sphinx-multiversion
 sphinx-rtd-theme
 sphinx-tabs
 sphinxcontrib-mermaid
+sphinx_tamer
```
Does not matter? seems it working without any problem?

```suggestion
sphinx-tamer
```

sloretz replied:
The underscore works because `.` and `_` get converted to `-` in Python project names.

https://peps.python.org/pep-0503/#normalized-names

fujitatomoya commented on .sphinx_tamer.yaml:
```diff
@@ -0,0 +1,10 @@
+    - nav2.org
+    - ros2.repos
+    - 2.X
+    - (\s|^)(\.)(msg|srv|action|NET|rviz)
```
```suggestion
    - (\s|^)(\.)(idl|msg|srv|action|NET|rviz)
```

fujitatomoya reviewed (approved):
a couple of comments.

sloretz committed:
Use normalized project name sphinx-tamer

Co-authored-by: Tomoya Fujita <Tomoya.Fujita@sony.com>
Signed-off-by: Shane Loretz <shane.loretz@gmail.com>

sloretz committed:
Make sentence checker ignore `.idl`

Co-authored-by: Tomoya Fujita <Tomoya.Fujita@sony.com>
Signed-off-by: Shane Loretz <shane.loretz@gmail.com>

fujitatomoya commented:
MERGING 🚀🚀🚀🚀🚀 @DLu thanks for amazing work 👍
